### ISSUE TITLE
Proposal: Enhance component styles prior to rendering

### ISSUE THREAD

**Issue by nevir** [rofrischmann/react-look#272]
**Version**: 1.0.0 (beta8)
**Environment**: Both
**Type**: Feature

-------------------
### Description

Currently, React Look performs much of its magic during component render, which has some downsides:

* React Look has to crawl all children of a component, which is a performance hot spot, especially for components that get frequently re-rendered.
* [Requires that we pass "malformed" props](https://github.com/rofrischmann/react-look/issues/216#issuecomment-208329927) (from React's perspective), which trips up prop type validation.
* Probably has some other downsides that I'm missing.

I'd like to try to address these issues with an API change to the component enhancer, by calculating styles _before_ rendering occurs, and exposing those calculated styles to the component during render.

The downside of this approach is that it prevents the use of mixins like `beforeAfter`, `empty`, etc.  (Or at least, those need to continue to be render-time plugins, if they are to be supported)

### Proposed API

Interestingly, this approach can also create `StyleSheet` objects on behalf of the caller; saves a bit of typing.

#### Stateless Components

I think the clearest approach is to inject `styles` as a prop:

```jsx
function Button({children, styles}) {
  return <div className={styles.root}>{children}</div>;
}

const styles = {
  root: {
    borderRadius: 2,
    backgroundColor: '#00D8FF',
    color: 'white',
    'disabled=true': {
      backgroundColor: '#CCC',
    },
  },
};

export default look(styles)(Button);
```

Alternatively, the calculated styles could be injected as a positional argument.  It gets cumbersome because functional components already receive two arguments - `(props, context)`.  Adding a third ends up being pretty verbose (`(props, _context, styles)`).

#### ES2015 classes

ES2015 classes are pretty straightforward; we can just expose the styles via data on the instance.  Styles stored directly on the instance seem best: we can't mutate props, and we shouldn't assume that component state exists, or is of a particular shape.

```jsx
const styles = {
  root: {
    borderRadius: 2,
    backgroundColor: '#00D8FF',
    color: 'white',
    'disabled=true': {
      backgroundColor: '#CCC',
    },
  },
};

@look(styles)
export default class Button extends React.Component {
  render() {
    return <div className={this.styles.root}>{this.props.children}</div>;
  }
}
```

**Comment by nevir** [comment]
I'm currently hacking on this to validate the approach, as well

We are using our own home rolled style management library that does something similar to this; it's worked very well for us so far (in production).  I think the approach will work well for React Look, too

**Comment by nevir** [comment]
The downside of this approach is that it prevents the use of mixins like `beforeAfter`, `empty`, etc.  (Or at least, those need to be a render-time plugin, if they are to be supported)

**Comment by rofrischmann** [comment]
Thanks for your proposal :) I will consider it wisely when implementing a more powerful rendering mechanism.

**Comment by rofrischmann** [comment]
I will close this one as I got a general issue to improve the render mechanism, this issue still will be linked there to not forget about this idea when getting serious about implementations!
Yet I want to point out, that Look already prerenders your styles and directly returns classNames. The fact that it wraps the render method and sometimes adds styles is because there are (as you also pointed out) dynamic plugins that need to be resolved during render.
Your proposal somehow would be implemented by just leaving out the dynamic aspect and not wrapping the render process anymore. In my opinion those dynamic styles are the only reason why one should use Look instead of simple CSS lately. But also you made me brainstorming last time I was running and I found an approach which might keep the benefits of both ideas. Keep updated soon!
